apps/www/_blog/2023-04-14-pg-tle.mdx

```diff
@@ -0,0 +1,66 @@
+
+# Introduction
+
+Before exploring what a Trusted Language Extension is, let’s cover the basics of Postgres Extensions and the notion of “Trusted Languages” in general.
+
+A Postgres Extension is a bundle of SQL and other code, primarily [Data Definition Language](https://www.postgresql.org/docs/current/ddl.html) (DDL) code installed into the server’s filesystem, and instantiated in a database with the `[CREATE EXTENSION](https://www.postgresql.org/docs/current/sql-createextension.html)` command. This code can contain any SQL, but almost always includes additional function objects created with the `[CREATE FUNCTION](https://www.postgresql.org/docs/current/sql-createfunction.html)` DDL statement that expose the core features of the extension to the database in some programming language, including an external library function written in C.
+
+Extensions provide a wide variety of additional functionality to Postgres and allow third party feature distribution, like new data types, storage engines, helper functions, even entire language runtimes like Javascript or Python, without requiring changes to the Postgres database itself.
+
+Some examples of popular Postgres extensions include:
+
+- [PostGIS](https://postgis.net/): for working with geospatial data
+- [Timescale DB](https://www.timescale.com/): for working with timeseries data
+- [Citus](https://www.citusdata.com/): for working with distributed tables
+- [plv8](https://plv8.github.io/): for running Node.js inside Postgres
+- [pgsodium](https://github.com/michelp/pgsodium): for accessing the libsodium encryption API
+
+Before extensions can be created, they must first be installed on the filesystem of the database server, which is an important security measure because extensions in some cases may require external libraries and can interact with the internals of Postgres. Their installation is something that requires careful consideration.
+
+## Trusted vs Untrusted
+
+As mentioned, extensions often include function objects created with `CREATE FUNCTION`. Many extensions provide functions whose implementation is written in C, and creating them in a database means that the compiled C code is “dynamically linked” into your running Postgres process. These dynamically-loaded libraries can now access every aspect of your running database process, right down to raw memory. They are essentially database superusers on steroids. Because of this, C is an “untrusted language” and installing extensions written in C requires filesystem access.
```

Comment:
```suggestion
Many extensions provide functions whose implementation is written in C, and creating them in a database means that the compiled C code is “dynamically linked” into your running Postgres process. These dynamically-loaded libraries can now access every aspect of your running database process, right down to raw memory. They are essentially database superusers on steroids. Because of this, C is an “untrusted language” and installing extensions written in C requires filesystem access.
```